File: cpp/gms/gms-filters-provider/src/parameters/IIRFilterParameters.cpp
#include "IIRFilterParameters.hh"

IIRFilterParameters::IIRFilterParameters(int groupDelaySec,
                                         bool isDesigned,
                                         double sampleRateHz,
                                         double sampleRateToleranceHz,
                                         std::vector<double> sosNumerator,
                                         std::vector<double> sosDenominator,
                                         std::vector<double> sosCoefficients) : sosNumerator(sosNumerator),
                                                                                sosDenominator(sosDenominator),
                                                                                sosCoefficients(sosCoefficients),
                                                                                BaseFilterParameters(groupDelaySec, isDesigned, sampleRateHz, sampleRateToleranceHz){
                                                                                    this->numberOfSos = sosNumerator.size() / 3;
                                                                                };

#if (__EMSCRIPTEN__)

emscripten::val IIRFilterParameters::getSosNumeratorAsTypedArray()
{
  return emscripten::val(emscripten::typed_memory_view(sosNumerator.size(), sosNumerator.data()));
}

emscripten::val IIRFilterParameters::getSosDenominatorAsTypedArray()
{
  return emscripten::val(emscripten::typed_memory_view(sosDenominator.size(), sosDenominator.data()));
}

emscripten::val IIRFilterParameters::getSosCoefficientsAsTypedArray()
{
  return emscripten::val(emscripten::typed_memory_view(sosCoefficients.size(), sosCoefficients.data()));
}

#endif

File: typescript/user-interface/packages/ui-wasm/src/ts/gms-filters-provider/gms-filters/types/iir-filter-parameters.ts
import type { VectorDouble } from '../../../types/vector-double';
import type { WasmVectorDouble } from './vector-double';

export interface WasmIIRFilterParametersModule {
  new (
    groupDelaySec: number,
    isDesigned: boolean,
    sampleRateHz: number,
    sampleRateToleranceHz: number,
    sosNumerator: WasmVectorDouble,
    sosDenominator: WasmVectorDouble,
    sosCoefficients: WasmVectorDouble
  ): WasmIIRFilterParameters;
}

export interface WasmIIRFilterParameters {
  $$: {
    ptr: number;
  };
  delete: () => void;
  isDesigned: boolean;
  groupDelaySec: number;
  sampleRateHz: number;
  sampleRateToleranceHz: number;
  sosNumerator: VectorDouble;
  sosDenominator: VectorDouble;
  sosCoefficients: VectorDouble;
  getSosNumeratorAsTypedArray: () => number[];
  getSosDenominatorAsTypedArray: () => number[];
  getSosCoefficientsAsTypedArray: () => number[];
}
